cirq/optimizers/three_qubit_decomposition.py

```diff
@@ -0,0 +1,272 @@
+    q1: ops.Qid,
+    q2: ops.Qid,
+    u1: np.ndarray,
+    u2: np.ndarray,
+    shift_left: bool = True,
+    diagonal: np.ndarray = np.eye(4),
+    atol: float = 1e-8,
+):
+    r"""Converts a two qubit double multiplexor to circuit.
+    Input: u1 ⊕ u2, with select qubit a (i.e. a = |0> => u1(b,c),
+    a = |1> => u2(b,c).
+
+    We want this:
+        $$
+        u1 ⊕ u2 = v ⊕ v @ D ⊕ D^{\dagger} @ W ⊕ W
+        $$
+    We can get it via:
+        $$
+        u1 = v @ D @ W       (1)
+        u2 = v @ D^{\dagger} @ W (2)
+        $$
+
+    We can derive
+        $$
+        u1u2^{adj}= v @ D^2 @ v^{\dagger}, (3)
+        $$
+
+    i.e the eigendecomposition of $u1u2^{\dagger}$ will give us D and v.
+    W is easy to derive from (2).
+
+    This function, after calculating v, D and W, also returns the circuit that
+    implements these unitaries: v, W on qubits b, c and the middle diagonal
+    multiplexer on a,b,c qubits.
+
+    The resulting circuit will have only known two-qubit and one-qubit gates,
+    namely CZ, CNOT and rx, ry, PhasedXPow gates.
+
+    Args:
+        q0: first qubit
+        q1: second qubit
+        q2: third qubit
+        u1: two-qubit operation on b,c for a = |0>
+        u2: two-qubit operation on b,c for a = |1>
+        shift_left: return the extracted diagonal or not
+        diagonal: an incoming diagonal to be merged with
+    Returns:
+        The circuit implementing the two qubit multiplexor consisting only of
+        known two-qubit and single qubit gates
+    """
+    u1u2 = u1 @ u2.conj().T
+    eigvals, v = cirq.unitary_eig(u1u2)
+    d = np.diag(np.sqrt(eigvals))
+
+    w = d @ v.conj().T @ u2
+
+    circuit_u1u2_mid = _middle_multiplexor_to_ops(q0, q1, q2, eigvals)
+
+    v = diagonal @ v
+
+    d_v, circuit_u1u2_r = opt.two_qubit_matrix_to_diagonal_and_operations(q1, q2, v, atol=atol)
+
+    w = d_v @ w
+
+    # if it's interesting to extract the diagonal then let's do it
+    if shift_left:
+        d_w, circuit_u1u2_l = opt.two_qubit_matrix_to_diagonal_and_operations(q1, q2, w, atol=atol)
+    # if we are at the end of the circuit, then just fall back to KAK
+    else:
+        d_w = np.eye(4)
+        circuit_u1u2_l = opt.two_qubit_matrix_to_operations(
+            q1, q2, w, allow_partial_czs=False, atol=atol
+        )
+
+    return d_w, cirq.Circuit([circuit_u1u2_l, circuit_u1u2_mid, circuit_u1u2_r])
+
+
+def _middle_multiplexor_to_ops(q0: ops.Qid, q1: ops.Qid, q2: ops.Qid, eigvals: np.ndarray):
+    theta = np.real(np.log(np.sqrt(eigvals)) * 1j * 2)
+    angles = _multiplexed_angles(theta)
+    rzs = [cirq.rz(angle).on(q0) for angle in angles]
+    ops = [
+        rzs[0],
+        cirq.CNOT(q1, q0),
+        rzs[1],
+        cirq.CNOT(q2, q0),
+        rzs[2],
+        cirq.CNOT(q1, q0),
+        rzs[3],
+        cirq.CNOT(q2, q0),
+    ]
+    return _optimize_multiplexed_angles_circuit(ops)
+
+
+def _multiplexed_angles(theta: Union[List[float], np.ndarray]) -> np.ndarray:
```

Comment:
I think I'll push back on this one. I tried it and I'm not convinced - as while it might save some space, I'm not sure it will be more readable / understandable / testable. I really like the fact that the individual building blocks map to the concepts in the paper. I find that easier to follow. Also, the unit tests are thoroughly testing each primitive one by one. If I start to merge them further up, I'd have to write much more complicated tests.

Reply:
Fair enough.

Regarding tests: Unit tests that verify behavior of private help functions make refactoring harder. Ideally, tests would verify behavior of such helper functions via specially chosen input data. This is why many best practices docs (e.g. Google's own testing best practices) encourage testing public APIs.